File: Sprint7/E/C++/main.cpp
// https://contest.yandex.ru/contest/25596/problems/E/
#include <iostream>
#include <vector>
#include <algorithm>

const int DP_MAX = 1e5;

int main() {
    int x, k;
    std::cin >> x >> k;

    std::vector<int> denominations(k);
    for (int i = 0; i < k; i++) {
        std::cin >> denominations[i];
    }

    std::vector<int> dp(x + 1, DP_MAX);
    dp[0] = 0;

    for (int i = 1; i <= x; i++) {
        for (int d : denominations) {
            if (i - d >= 0) {
                dp[i] = std::min(dp[i], dp[i - d] + 1);
            }
        }
    }

    if (dp[x] == DP_MAX) {
        std::cout << -1 << std::endl;
    } else {
        std::cout << dp[x] << std::endl;
    }

    return 0;
}

File: Sprint7/E/Rust/src/main.rs
// https://contest.yandex.ru/contest/25596/problems/E/

const DP_MAX: usize = 100_000;

fn main() {
    let mut lines = std::io::stdin().lines();
    let x = lines.next().unwrap().unwrap().parse::<usize>().unwrap();
    let _k = lines.next().unwrap().unwrap().parse::<usize>().unwrap();

    let denominations = lines
        .next()
        .unwrap()
        .unwrap()
        .split_whitespace()
        .map(|x| x.parse::<usize>().unwrap())
        .collect::<Vec<usize>>();

    let mut dp = vec![DP_MAX; x + 1];
    dp[0] = 0;

    for i in 1..(x + 1) {
        for d in denominations.iter() {
            if i as i64 - *d as i64 >= 0 {
                dp[i] = std::cmp::min(dp[i], dp[i - *d] + 1);
            }
        }
    }

    if dp[x] == DP_MAX {
        println!("-1");
    } else {
        println!("{}", dp[x]);
    }
}
